src/subhalo_tracking.cpp

```diff
@@ -175,30 +183,187 @@ inline HBTInt GetLocalHostId(HBTInt pid, const HaloSnapshot_t &halo_snap, const
+   * bound was the particle is */
+  for (int BoundRanking = 0; BoundRanking < HBTConfig.NumTracerHostFinding; BoundRanking++)
+  {
+    // Not a valid tracer (NOTE: should we break or continue here?)
+    if (particle_hosts[BoundRanking] == -2)
+      continue;
+
+    /* If the key is not present, the score gets zero-initialised by default */
+    CandidateHosts[particle_hosts[BoundRanking]] += 1.0 / (1 + pow(float(BoundRanking), 0.5));
+  }
+
+  /* Select candidate host with the highest score. */
+  HBTInt HostId = -2; // Default value
+  float MaximumScore = 0;
+
+  for (auto candidate : CandidateHosts)
+  {
+    if (candidate.second > MaximumScore)
+    {
+      HostId = candidate.first;
+      MaximumScore = candidate.second;
+    }
+  }
+
+  return HostId;
+}
+
+/* Assign a host to a given subgroup based on the FoF membership of the
+ * NumTracerHostFinding most bound collisionless particles. This decision is
+ * weighted by the binding energy ranking the particles had in the previous
+ * output. Used during search across multiple tasks.*/
+IdRank_t DecideLocalHostId(vector<IdRank_t>::const_iterator particle_hosts)
+{
+  /* To store unique host candidates, their rank, and the matching score. */
+  unordered_map<HBTInt, float> CandidateHosts, CandidateHostRank;
+
+  /* Iterate over the particle list, and weight each candidate score by how
+   * bound was the particle is */
+  for (int BoundRanking = 0; BoundRanking < HBTConfig.NumTracerHostFinding; BoundRanking++)
+  {
+    // Not a valid tracer (NOTE: should we break or continue here?)
+    if (particle_hosts[BoundRanking].Id == -2)
+      continue;
+
+    /* If the key is not present, the score gets zero-initialised by default */
+    CandidateHosts[particle_hosts[BoundRanking].Id] += 1.0 / (1 + pow(float(BoundRanking), 0.5));
+
+    /* If we have hostless particles, we assign the rank to be the one where the
+     * most bound hostless particle is present. */
+    if (particle_hosts[BoundRanking].Id == -1)
+    {
+      auto findit = CandidateHostRank.find(particle_hosts[BoundRanking].Id);
+      if (findit == CandidateHostRank.end()) // First time we find a hostless particle.
+        CandidateHostRank[particle_hosts[BoundRanking].Id] = particle_hosts[BoundRanking].Rank;
+    }
+
+    // Store its rank
+    CandidateHostRank[particle_hosts[BoundRanking].Id] = particle_hosts[BoundRanking].Rank;
+  }
+
+  /* Select candidate host with the highest score. */
+  IdRank_t HostIdRank{-2, -1};
+  float MaximumScore = 0;
+
+  for (auto candidate : CandidateHosts)
+  {
+    if (candidate.second > MaximumScore)
+    {
+      HostIdRank.Id = candidate.first;
+      MaximumScore = candidate.second;
+    }
+  }
+
+  /* Assign target rank where the subhalo will be moved to. */
+  HostIdRank.Rank = CandidateHostRank[HostIdRank.Id];
+
+  return HostIdRank;
+}
+
 void FindLocalHosts(const HaloSnapshot_t &halo_snap, const ParticleSnapshot_t &part_snap, vector<Subhalo_t> &Subhalos,
```

Comment:
We can either use John's new functions to retrieve information about arbitrary particle IDs, or make use of the updated information that subhalo particles have to prevent communication. The last option was not possible before, since LoadSnapshot did not provide FOF group Ids, but the #26 now makes that information available.

Reply:
Are you suggesting that as an optimization for AssignHosts? If so, maybe that can be a separate pull request?

Reply:
Yup! As you say, we can do it in a separate request. At the moment, that is the second most expensive step after I/O in FLAMINGO, right? 

It should only involve iterating over the Particles vector of each resolved subhalo until we get the 10 most bound tracers, omitting the need for MPI communication.